rfd/0019-cluster-routing.md

```diff
@@ -0,0 +1,36 @@
+---
+authors: Russell Jones (rjones@goteleport.com)
+state: draft
+---
+
+# RFD 19 - Cluster Routing
+
+## What
+
+This RFD proposes changes to how Teleport issues certificates and routes requests within clusters. These changes would make using certificate issues by a root cluster directly on leaf clusters substantially easier.
+
+## Why
+
+At the moment connections to a leaf clusters typically go through a root cluster. If the latency to the root cluster is high, accessing servers within a leaf cluster leads to a frustrating user experience. Allowing direct use of certificates issues by a root cluster on a leaf cluster mitigates this problem.
+
+Teleport supports this use case.
+
+Users can run `tsh login clusterName` to obtain a certificate with `RouteToCluster` set to a leaf cluster then use `tsh ssh -J clusterName severName` to bypass the root cluster. However, switching between multiple leaf clusters is cumbersome as `tsh login clusterName` has to be issued each time the user wishes to switch between clusters and it also makes working on multiple clusters concurrently almost impossible.
+
+## Details
+
+To support this use case, this RFD proposes `tsh login` behavior be changed to issue certificates with an empty `RouteToCluster` field. An empty `RouteToCluster` field on a certificate would indicate to a cluster that the request should be routed locally baring any other routing directives.
+ 
+This would allow users to use `tsh ssh -J clusterName serverName` and work on multiple leaf clusters concurrently without having to obtain a new certificate. When these requests arrive on the leaf cluster, the leaf cluster would route them locally instead of attempting (and failing) to route them to the root cluster.
+
+No changes would be made to role mapping or to the behavior when the `RouteToCluster` field is set.
+
+### Considerations
```

Comment:
A few things to address:
- how does this affect cluster labels and RBAC?
- how does role mapping happen? I assume the leaf checks the cert signer, notices that it's a remote but trusted cluster and applies its mapping. But I'm not 100% sure it works this way and it's a very obscure detail that most readers won't understand.
- are there any other security controls that are enforced at the root cluster that we'd lose? some parts of audit log or concurrent session limiting maybe?

Reply:
RBAC is indeed a big problem with this proposal. Currently, the necessity of calling `tsh login clusterName` restricts `RouteToCluster` to clusters the user has a right to access. In addition, any attempt to use the jumphost with a non-matching `RouteToCluster` would result in a "cluster not found" error.

With empty `RouteToCluster` interpreted as local routing without any qualification, it becomes possible to access nodes that are supposed to be unavailable to the user, with the only limiting factors being the list of logins in the certificate and the knowledge of a node's address.

Reply:
@awly @andrejtokarcik Valid point about cluster label bypass.

What do you two think about the following alternative: We teach `tsh ssh -J` to request a certificate re-issue for the target cluster if the proxy address does not match the root proxy address (which should be saved in the profile)?

In that situation the first step would be to go to the proxy, loop over all remote clusters and check if the public address matches and of the leaf clusters and if so re-issue the certificate with a valid `RouteToCluster`. We could cache certificates at the proxy to prevent re-issue performance hit, this is what we do on the recording proxy.

Feels like a safer fix as well since changed are fairly limited.

Reply:
Sounds good. It's a much more isolated and safe change.

`tsh` should probably cache certs on the client side, just like we do with k8s and db access.

Reply:
@awly Yeah, I was thinking about but didn't want to clutter the keys directory, didn't realize we were already caching for Kuberentes and Database Access, in that case it's probably better to client side caching.